### PR TITLE
Breakpoint

### PR DESCRIPTION
**Pull request by arnabcs17b006**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (Wenzel) [inline .vscode/settings.json:1]: we don't need your IDE files

File: src/lib.rs [hunk 2]
```diff
@@ -109,12 +269,81 @@ impl XenControl {
         last_error!(self, ())
     }
 
+    pub fn monitor_software_breakpoint(&self, domid: u32, enable: bool) -> Result<()> {
+        let xc = self.handle.as_ptr();
+        (self.libxenctrl.clear_last_error)(xc);
+        let rc = (self.libxenctrl.monitor_software_breakpoint)(xc, domid, enable);
+        if rc < 0 {
+            println!("last OS error: {:?}", Error::last_os_error());
```
Comment (Wenzel) [inline src/lib.rs:277]: no println

**Review by Wenzel — CHANGES_REQUESTED** [review]

**Comment by Wenzel** [comment]
I don't get it either.
I looked at the generated bindings, and the `value` field is there
These are the bindings generated directly from xenvmevent-sys crate:
Capture d’écran de 2020-08-29 00-02-28

And surprise, this is the ones when xenctrl is built:
Capture d’écran de 2020-08-29 00-07-05

**Comment by arnabcs17b006** [comment]
It also works fine locally..

**Comment by Wenzel** [comment]
When I look at my `vm_event.h` header, i have this:
Capture d’écran de 2020-08-29 00-14-50ù

So I suspect this is linked to the version of Xen you are using.
I have Xen 4.11

Looking at the Xen source code, this [commit](https://github.com/xen-project/xen/commit/b55f6c299d316b00f65b63137fde8c2524fb638a#diff-4f1f0664b6782533c6ae2d860fb6e27fR263) introduces the old MSR value and changed the fields, likely since Xen 4.11 rc1

**Comment by Wenzel** [comment]
Also, about Ubuntu 16.04 failing on the CI, same thing
this [commit](https://github.com/xen-project/xen/commit/94db5dbb6c5f331dac35a8723a5d982ad17536d6) added the `insn_length` field, since Xen 4.8.0 rc1.

How should we handle backward compatibility with bindgen, since it generates the bindgen for the current headers only ?
cc @rageagainsthepc

**Comment by arnabcs17b006** [comment]
I thought the issue was with insn_len of vm_event_debug struct?

**Comment by Wenzel** [comment]
It is, this is exactly what my last comment was saying.

**Comment by Wenzel** [comment]
One solution: move to Xen 4.11 APIs, drop Ubuntu 16.04 compat ?

If you look at LibVMI, they define multiple `process_request_x` for each API version:
Here, the lowest, with only the gfn field available.
https://github.com/libvmi/libvmi/blob/master/libvmi/driver/xen/xen_events.c#L1384

**Comment by Wenzel** [comment]
The solution I see here:
drop Xen 4.6 support (Ubuntu 16.04), and only use Xen 4.9 (Ubuntu 18.04) from now on.

**Comment by rageagainsthepc** [comment]
We should also mention somewhere that we don't support Xen versions below 4.9.

**Comment by Wenzel** [comment]
https://github.com/Wenzel/xenctrl/pull/26 has been merged.
Arnab you can rebase and the CI should go green.

**Pull request closed without merging**